### PR TITLE
P lexers

### LINKED ISSUE #215: Port missing P lexers
Port missing `p` lexers to Chroma.

- Pan
- ParaSail
- Pawn
- PEG
- Perl6
- PHP
- Pike
- Pointless
- PostgreSQL console (psql)
- POVRay
- PowerShell Session
- Praat
- Prolog
- Properties
- PsySH console session for PHP
- PyPy Log
- Python
- Python 2.x
- Python 2.x Traceback
- Python console session
- Python Traceback

### PR DESCRIPTION
**Pull request by gandarez**
Port missing `p` lexers to Chroma.

- Pan
- ParaSail
- PEG
- Perl6
- Pike - Text analysis inherits from C++. However the include syntax seems to always have file extension which is not validated at C++ text analysis. [doc](https://pike.lysator.liu.se/generated/manual/modref/ex/cpp_3A_3A/_23include.html)
- Pointless
- PostgreSQL console (psql)
- PowerShell Session
- Praat
- Properties
- PsySH console session for PHP
- PyPy Log
- Python - Original code for Python3 ([here](https://github.com/pygments/pygments/blob/a590ac5ea7c00a41e253834306bfa19e38349c0b/pygments%2Flexers%2Fpython.py#L28)) was splitted into two files at Chroma [here](https://github.com/alecthomas/chroma/blob/master/lexers/p/python.go) and [here](https://github.com/alecthomas/chroma/blob/master/lexers/p/python3.go). The analyse text function has been ported only for `Python.go` and not `Python3.go`. Do you think I should replicate the same code for both files?
- Python 2.x
- Python 2.x Traceback
- Python console session
- Python Traceback

Closes https://github.com/wakatime/wakatime-cli/issues/215

### CODE REVIEW

File: lexers/p/python.go [hunk 1]
```diff
@@ -153,4 +154,10 @@ var Python = internal.Register(MustNewLexer(
 			{`\n`, LiteralStringSingle, nil},
 		},
 	},
-))
+).SetAnalyser(func(text string) float32 {
+	if matched, _ := shebang.MatchString(text, `pythonw?(3(\.\d)?)?`); matched {
```
Comment (dron22) [inline lexers/p/python.go]: The pygments implementation has a second condition checking for string `import ` at the beginning of the text: https://github.com/pygments/pygments/blob/master/pygments/lexers/python.py#L382

File: lexers/p/python2.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+		"root": {},
+	},
+).SetAnalyser(func(text string) float32 {
+	matched, _ := shebang.MatchString(text, `pythonw?2(\.\d)?`)
+
+	if len(text) > 1000 {
+		text = text[:1000]
+	}
+
+	if matched || strings.Contains(text, "import ") {
```
Comment (dron22) [inline lexers/p/python2.go]: I think you accidently mixed this up. The check for `import ` is used in `PythonLexer`, while `Python2Lexer` does not have this in pygments: https://github.com/pygments/pygments/blob/master/pygments/lexers/python.py#L606

File: lexers/p/python2_test.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/alecthomas/assert"
+	"github.com/alecthomas/chroma"
+	"github.com/alecthomas/chroma/lexers/p"
+)
+
+func TestCpp_AnalyseText(t *testing.T) {
```
Comment (dron22) [inline lexers/p/python2_test.go]: small copy&paster leftover
```suggestion
func TestPython2_AnalyseText(t *testing.T) {
```

File: lexers/p/python_test.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/alecthomas/assert"
+	"github.com/alecthomas/chroma"
+	"github.com/alecthomas/chroma/lexers/p"
+)
+
+func TestPython3_AnalyseText(t *testing.T) {
```
Comment (dron22) [inline lexers/p/python_test.go]: As lexer is named `Python`, let's have the test name aligned here. Pip finally ending support for Python2. Python3 is the new standard eventually. This took some years :joy: 
```suggestion
func TestPython_AnalyseText(t *testing.T) {
```

File: lexers/p/testdata/powershellsession.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package p
+
+import (
+	. "github.com/alecthomas/chroma" // nolint
+	"github.com/alecthomas/chroma/lexers/internal"
+)
+
+// PowerShellSession lexer.
+var PowerShellSession = internal.Register(MustNewLexer(
```
Comment (dron22) [inline lexers/p/testdata/powershellsession.go]: This lexer file accidently ended up in `testdata` folder.

File: lexers/p/perl6.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+		result /= 2
+	}
+
+	return result
+}))
+
+func perl6StripPod(text string) []string {
+	var (
+		inPod         bool
+		strippedLines []string = []string{}
```
Comment (dron22) [inline lexers/p/perl6.go]: You do not need to explicitly initialize the slice. This is only needed for maps, where you have to use `make()`
```suggestion
		strippedLines []string
```

File: lexers/p/perl6.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+		podRe                  = regexp.MustCompile(`^=(?:end|cut)`)
+		otherPodRe             = regexp.MustCompile(`^=\w+`)
```
Comment (dron22) [inline lexers/p/perl6.go]: Let's move all regex definitions to the top of the file. This way, they will be executed upon import of the file and any errors will be catched right away.

File: lexers/p/perl6.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+			inPod = false
+			continue
+		}
+
+		if otherPodRe.MatchString(line) {
+			inPod = true
+			continue
+		}
+
+		if inPod == false {
```
Comment (dron22) [inline lexers/p/perl6.go]: Let's shorten this
```suggestion
		if !inPod {
```

File: lexers/p/perl6.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	return strippedLines
+}
+
+func perl6GetSubgroups(match []string) map[string]string {
+	groups := make(map[string]string)
+
+	for i, name := range perl6AnalyserDeclScope.SubexpNames() {
+		if i > 0 && i <= len(match) {
```
Comment (dron22) [inline lexers/p/perl6.go]: I guess length of `SubexpNames()` slice and length of matches will always be the same. But if you want to keep the second condition, it should be changed to check for `i < len(match)`.
````suggestion
		if i > 0 && i < len(match) {
```

File: lexers/p/perl6_test.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+		},
+		"scoped class": {
+			Filepath: "testdata/perl6_scoped_class.pl6",
+			Expected: 1.0,
+		},
+		"assignment": {
+			Filepath: "testdata/perl6_assign.pl6",
+			Expected: 0.4,
+		},
+	}
```
Comment (dron22) [inline lexers/p/perl6_test.go:41]: Can we add another test case to test this pod removal?

**Review by dron22 — CHANGES_REQUESTED** [review]

File: lexers/p/python.go (hunk 1 above)
Comment (gandarez) [reply lexers/p/python.go]: my master branch was outdated.

File: lexers/p/python_test.go (hunk 4 above)
Comment (gandarez) [reply lexers/p/python_test.go]: harder decision to make!

File: lexers/p/python2.go (hunk 2 above)
Comment (gandarez) [reply lexers/p/python2.go]: for sure lol! tks.

File: lexers/p/testdata/powershellsession.go (hunk 5 above)
Comment (gandarez) [reply lexers/p/testdata/powershellsession.go]: OMG!

File: lexers/p/perl6_test.go (hunk 10 above)
Comment (gandarez) [reply lexers/p/perl6_test.go:41]: make sense 👍

File: lexers/p/perl6.go (hunk 7 above)
Comment (gandarez) [reply lexers/p/perl6.go]: I also renamed the variables to provide more context.

**Comment by gandarez** [comment]
@dron22 all changes applied thanks for the tips 👍

File: lexers/p/perl6.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+		if perl6AnalyserShell.MatchString(line) {
+			continue
+		}
+
+		// Match v6; use v6; use v6.0; use v6.0.0.
+		if perl6AnalyserV6.MatchString(line) {
+			return 1.0
+		}
+
+		//Match class, module, role, enum, grammar declarations.
```
Comment (dron22) [inline lexers/p/perl6.go]: missing blank space
```suggestion
		// Match class, module, role, enum, grammar declarations.
```

File: lexers/p/perl6_test.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+			Filepath: "testdata/perl6_scoped_class.pl6",
+			Expected: 1.0,
+		},
+		"assignment": {
+			Filepath: "testdata/perl6_assign.pl6",
+			Expected: 0.4,
+		},
+		"strip pod": {
+			Filepath: "testdata/perl6_pod.pl6",
+			Expected: 0,
```
Comment (dron22) [inline lexers/p/perl6_test.go]: What I had in mind was adding a test, where pod is removed, but the result is > 0. To check that it correctly removed the pod section and nothing else.

**Review by dron22 — APPROVED** [review]
Two small last remarks. Approving already :+1:

**Pull request merged**